File: ch08/radix_sort.js


const extract_b = function(n, digit) {
    let p1 = 10 ** digit;
    let p2 = 10 ** (digit - 1);
    return ((n % p1) - (n % p2)) / p2;
}

const extract = function(n, digit) {
    if (n < 0) {
        throw new Error("n must be a positive integer");
    } else if (digit < 1) {
        throw new Error("digit at least 1");
    } else {
        return extract_b(n, digit);
    }
}

const counting_stable = function(A, d) {
    let record = new Array();
    for (let i = 0; i != 10; i += 1) {
        record.push(0);
    }

    for (let i = 0; i != A.length; i += 1) {
        let n = extract(A[i], d);
        record[n] += 1;
    }

    for (let i = 1; i != record.length; i += 1) {
        record[i] = record[i] + record[i - 1];
    }

    let out = new Array(A.length);
    for (let i = A.length - 1; i >= 0; i -= 1) {
        let n = extract(A[i], d);
        let index = record[n] - 1;
        out[index] = A[i];
        record[n] = index;
    }

    return out;
}

const radix_sort = function(A, digits) {
    let temp = A;
    for (let i = 1; i <= digits; i += 1) {
        temp = counting_stable(temp, i);
    }
    array_copy(temp, A);
}

const array_copy = function(src, des) {
    for (let i = 0; i != src.length; i += 1) {
        des[i] = src[i];
    }
}

//--------------------------------------

const randomInt = function(min, max) {
    return Math.floor(Math.random() * (max - min) + min);
}

const setRandomArray = function(size, min, max) {
    let v = new Array();
    for (let i = 0; i < size; i = i+1) {
        v.push(randomInt(min, max));
    }
    return v;
}

let count = 100;
let n_min = 10;
let n_max = 10000;
let digits = 4;

let A = setRandomArray(count, n_min, n_max);
console.log(A);
radix_sort(A, digits);
console.log(A);

File: ch08/radix_sort.py

import random


def extract_b(n, digit):
  p1 = 10 ** digit
  p2 = 10 ** (digit - 1)
  return ((n % p1) - (n % p2)) // p2


def extract(n, digit):
  if (n < 0):
    raise ValueError("extract: n must be a positive integer")
  elif (digit < 1):
    raise ValueError("extract: digit at least 1")
  else:
    return extract_b(n, digit)


def counting_stable(l, d):
  record = []
  for i in range(0, 10):
    record.append(0)
  for i in range(0, len(l)):
    n = extract(l[i], d)
    record[n] += 1

  for i in range(1, len(record)):
    record[i] = record[i] + record[i - 1]

  out = []
  for i in range(0, len(l)):
    out.append(0)
  for i in range(len(l) - 1, -1, -1):
    n = extract(l[i], d)
    index = record[n] - 1
    out[index] = l[i]
    record[n] = index

  return out


def radix_sort(l, digits):
  temp = l
  for i in range(1, digits + 1):
    temp = counting_stable(temp, i)

  list_copy(temp, l)


def list_copy(src, des):
  for i in range(0, len(src)):
    des[i] = src[i]

#---------------------------------------

def setRandomList(size, n_min, n_max):
  bound = n_max - n_min
  v = []
  for i in range(0, size):
    v.append(random.randint(0, bound) + n_min)
  return v

def printList(v):
  print('(List: #', end='')
  print(len(v), end='')
  print(' [', end='')
  for i in range(0, len(v)):
    print(v[i], end='')
    print(' ', end='')
  print('])')

count = 100
n_min = 10
n_max = 100000
digits = 5

A = setRandomList(count, n_min, n_max)
printList(A)
radix_sort(A, digits)
printList(A)
